CI Failure: org.elasticsearch.recovery.RelocationIT testIndexAndRelocateConcurrently

Unable to reproduce locally.

CI Failure:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos/852/consoleFull

Reproduce With:
gradle :core:integTest -Dtests.seed=4856A7A6357BA799 -Dtests.class=org.elasticsearch.recovery.RelocationIT -Dtests.method="testIndexAndRelocateConcurrently" -Dtests.security.manager=true -Dtests.locale=sk -Dtests.timezone=America/Argentina/Rio_Gallegos

The test encounters a local checkpoint mismatch, probably because the translog is missing an entry (replica cannot advance local checkpoint).

Relevant log lines:

```
   > Throwable #1: java.lang.AssertionError: Some expected ids were not found in search results: [ꩣꩺꩪ꩹ꩣ꩸ꩪꩿ꩸a180, եՍճՂմ֏ԽդՍa216, 𓇨𓏀𓆨𓈡aa238, ꓛꓵꓶꓘꓷꓑꓝꓼꓚa226, ㆴㆠㆳㆩㆢㆭㆾㆡㆠa137, 󠄱󠄼󠇗󠄝aa240, 🉢🉡🊪🈵aa126, 〚〄〼〻〃〳」〦〡a122, ᥃ᤔᤪ᤮ᤓ᥆᤿᥆᥉a149, 𐋘𐋒𐋀𐋟aa207, ꥧꥫꥧꥡꥴꥼꥹꥬꥳa206, ︉︁︌︆︀︊️︉︃a121, ╷╄┴┩╺┌┅╖╓a182]. Total shards: 3 Successful shards: 3 & 0 shard failures:
   > Expected: <0>
   >      but: was <13>
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertSearchHits(ElasticsearchAssertions.java:229)
   > 	at org.elasticsearch.recovery.RelocationIT.testIndexAndRelocateConcurrently(RelocationIT.java:512)
   > 	at java.lang.Thread.run(Thread.java:745)Throwable #2: java.lang.AssertionError: [test][0], node[g-azBk_eSQOcK9hsQcsYow], [R], s[STARTED], a[id=g7JRvvK8RAWPlZIKShHfIQ] local checkpoint mismatch
   > Expected: <73L>
   >      but: was <53L>
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.recovery.RelocationIT.lambda$beforeIndexDeletion$1(RelocationIT.java:132)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:711)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:685)
   > 	at org.elasticsearch.recovery.RelocationIT.beforeIndexDeletion(RelocationIT.java:116)
   > 	at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:579)
   > 	at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2058)
   > 	at java.lang.Thread.run(Thread.java:745)
   > 	Suppressed: java.lang.AssertionError: [test][0], node[g-azBk_eSQOcK9hsQcsYow], [R], s[STARTED], a[id=g7JRvvK8RAWPlZIKShHfIQ] local checkpoint mismatch
   > Expected: <73L>
   >      but: was <53L>
   > 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 		at org.elasticsearch.recovery.RelocationIT.lambda$beforeIndexDeletion$1(RelocationIT.java:132)
   > 		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:699)
   > 		... 38 more
```

I'll have to dig deeper what's causing this /cc: @jasontedor 
More recent failure: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=oraclelinux/883/consoleFull

I've closely examined the failure here:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=oraclelinux/883/consoleFull

What is happening here is the following. We have 4 nodes:

```
{node_t0}{L3CpmEfoTMuwIFXLNEy4yA}{olvHdPIXRI-g6hl46IOImA}{127.0.0.1}{127.0.0.1:9400}{color=blue}, master
{node_t1}{OJ_UPhr4S8GEepzK3zNR1Q}{TV4yTxtKRxa_9oF_s9yWBw}{127.0.0.1}{127.0.0.1:9401}{color=blue}
{node_t2}{poCFuyO6Ss6wqlv_amKDKw}{ZWhmdAHmT2uQ9xM1WxSOIA}{127.0.0.1}{127.0.0.1:9402}{color=red}
{node_t3}{cYkbR4D_TCeQ6CxzlUd_RQ}{bgCrxqWzQvOfBWyYNOXW9A}{127.0.0.1}{127.0.0.1:9403}{color=red}
```

and are in a state where we have a relocating primary (from node_t2 to node_t0) together with a relocating replica (from node_t3 to node_t1):

```
cluster state version: 14
  --------[test][2], node[poCFuyO6Ss6wqlv_amKDKw], relocating [L3CpmEfoTMuwIFXLNEy4yA], [P], s[RELOCATING], a[id=iAUBqd0rSbyA-B9-f8mbRQ, rId=xKA60_nPQaWDi-sh9NvGHg]
  --------[test][2], node[cYkbR4D_TCeQ6CxzlUd_RQ], relocating [OJ_UPhr4S8GEepzK3zNR1Q], [R], s[RELOCATING], a[id=jrBmMQUoRb6gi1LniVy-3w, rId=mcc4_uPbQi-qmxM9jApG5A]
```

Primary relocation completes before the replica relocation, which means that replica recovery is cancelled by changing the allocation id of the replica relocation target (see #23926):

```
cluster state version: 15
  --------[test][2], node[L3CpmEfoTMuwIFXLNEy4yA], [P], s[STARTED], a[id=xKA60_nPQaWDi-sh9NvGHg]
  --------[test][2], node[cYkbR4D_TCeQ6CxzlUd_RQ], relocating [OJ_UPhr4S8GEepzK3zNR1Q], [R], s[RELOCATING], a[id=jrBmMQUoRb6gi1LniVy-3w, rId=7JPzU857TwepFt0r0Xkkpw]
```

This cluster state is first applied on the node that has the replica relocation target, which then removes the shard as the allocation id changed, creates a new initializing shard and restarts recovery from the new primary:

```
[2017-04-25T15:10:00,165][DEBUG][o.e.i.c.IndicesClusterStateService] [node_t1] [test][2] removing shard (stale allocation id, stale [test][2], node[OJ_UPhr4S8GEepzK3zNR1Q], relocating [cYkbR4D_TCeQ6CxzlUd_RQ], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=mcc4_uPbQi-qmxM9jApG5A, rId=jrBmMQUoRb6gi1LniVy-3w], new [test][2], node[OJ_UPhr4S8GEepzK3zNR1Q], relocating [cYkbR4D_TCeQ6CxzlUd_RQ], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=7JPzU857TwepFt0r0Xkkpw, rId=jrBmMQUoRb6gi1LniVy-3w])

[2017-04-25T15:10:00,173][TRACE][o.e.i.r.PeerRecoveryTargetService] [node_t1] [test][2] started recovery from {node_t0}{L3CpmEfoTMuwIFXLNEy4yA}{olvHdPIXRI-g6hl46IOImA}{127.0.0.1}{127.0.0.1:9400}{color=blue}, id [448]

[2017-04-25T15:10:00,175][TRACE][o.e.i.r.PeerRecoveryTargetService] [node_t1] [test][2] preparing for sequence-number-based recovery starting at local checkpoint [0] from [{node_t0}{L3CpmEfoTMuwIFXLNEy4yA}{olvHdPIXRI-g6hl46IOImA}{127.0.0.1}{127.0.0.1:9400}{color=blue}]
```

The new primary goes ahead with the recovery using sequence numbers:

```
[2017-04-25T15:10:00,175][TRACE][o.e.i.r.PeerRecoverySourceService] [node_t0] [test][2] starting recovery to {node_t1}{OJ_UPhr4S8GEepzK3zNR1Q}{TV4yTxtKRxa_9oF_s9yWBw}{127.0.0.1}{127.0.0.1:9401}{color=blue}

[2017-04-25T15:10:00,176][TRACE][o.e.i.r.RecoverySourceHandler] [node_t0] [test][2][recover to node_t1] performing sequence numbers based recovery. starting at [0]
```

Note that in contrast to recoveries that go through phase 1 (and call acquireIndexCommit), sequence number based recoveries (afaics) can be done on a shard that is not marked as STARTED yet. The critical part here is that node_t0 which is the primary relocation target has not applied the cluster state yet where it knows that the relocation successfully completed, it's shard state is currently in POST_RECOVERY. Another critical aspect is that recovery validation only checks that the recovery source node knows that there is an initializing shard on the recovery target node but does not check whether that shard has the correct allocation id. This means that the recovery source node thinks in this case that the recovery target on node_t1 has allocation id `mcc4_uPbQi-qmxM9jApG5A` instead of `7JPzU857TwepFt0r0Xkkpw`. This means that other index requests that arrive on the primary after the recovery is started, are replicated to node_t1 by using the wrong allocation id. This leads to an allocation mismatch failure on the target node:

```
[2017-04-25T15:10:00,195][TRACE][o.e.a.b.TransportShardBulkAction] [node_t0] [[test][2]] failure while performing [indices:data/write/bulk[s]] on replica [test][2], node[OJ_UPhr4S8GEepzK3zNR1Q], relocating [cYkbR4D_TCeQ6CxzlUd_RQ], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=mcc4_uPbQi-qmxM9jApG5A, rId=jrBmMQUoRb6gi1LniVy-3w], request [BulkShardRequest [[test][2]] containing [index {[test][type1][ऺॺॻलु़ॊफॆa121], source[{"field1":"one hundred twenty-one "}]}]]
  1> org.elasticsearch.transport.RemoteTransportException: [node_t1][127.0.0.1:9401][indices:data/write/bulk[s][r]]
  1> Caused by: org.elasticsearch.index.shard.ShardNotFoundException: expected aID [mcc4_uPbQi-qmxM9jApG5A] but found [7JPzU857TwepFt0r0Xkkpw]
  1> 	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncReplicaAction.doRun(TransportReplicationAction.java:568) ~[main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
  1> 	at ...
```

As this is a `ShardNotFoundException`, we just ignore it on the sender site (i.e. don't fail the target shard). This means that that write is not properly replicated to the target shard, resulting in the missing write on the replica:

```
> Throwable #1: java.lang.AssertionError: Some expected ids were not found in search results: [ऺॺॻलु़ॊफॆa121]. Total shards: 10 Successful shards: 10 & 0 shard failures:
   > Expected: <0>
   >      but: was <1>
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertSearchHits(ElasticsearchAssertions.java:229)
   > 	at org.elasticsearch.recovery.RelocationIT.testIndexAndRelocateConcurrently(RelocationIT.java:512)
   > 	at java.lang.Thread.run(Thread.java:745)Throwable #2: java.lang.AssertionError: [test][2], node[OJ_UPhr4S8GEepzK3zNR1Q], [R], s[STARTED], a[id=7JPzU857TwepFt0r0Xkkpw] local checkpoint mismatch
   > Expected: <21L>
   >      but: was <15L>
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.recovery.RelocationIT.lambda$beforeIndexDeletion$1(RelocationIT.java:132)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:711)
   > 	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:685)
   > 	at org.elasticsearch.recovery.RelocationIT.beforeIndexDeletion(RelocationIT.java:116)
   > 	at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:579)
   > 	at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2057)
   > 	at java.lang.Thread.run(Thread.java:745)
   > 	Suppressed: java.lang.AssertionError: [test][2], node[OJ_UPhr4S8GEepzK3zNR1Q], [R], s[STARTED], a[id=7JPzU857TwepFt0r0Xkkpw] local checkpoint mismatch
   > Expected: <21L>
   >      but: was <15L>
   > 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 		at org.elasticsearch.recovery.RelocationIT.lambda$beforeIndexDeletion$1(RelocationIT.java:132)
   > 		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:699)
   > 		... 38 more
   > 	Suppressed: java.lang.AssertionError: [test][2], node[OJ_UPhr4S8GEepzK3zNR1Q], [R], s[STARTED], a[id=7JPzU857TwepFt0r0Xkkpw] local checkpoint mismatch
   > Expected: <21L>
   >      but: was <15L>
   > 		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 		at org.elasticsearch.recovery.RelocationIT.lambda$beforeIndexDeletion$1(RelocationIT.java:132)
   > 		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:699)
   > 		... 38 more
```

I've opened #24333 as a fix

--- a/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
@@ -84,7 +84,7 @@ public class PeerRecoverySourceService extends AbstractComponent implements Inde
         }
     }
 
-    private RecoveryResponse recover(final StartRecoveryRequest request) throws IOException {
+    private RecoveryResponse recover(StartRecoveryRequest request) throws IOException {
         final IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
         final IndexShard shard = indexService.getShard(request.shardId().id());
 
@@ -113,7 +113,22 @@ public class PeerRecoverySourceService extends AbstractComponent implements Inde
             throw new DelayRecoveryException("source node has the state of the target shard to be [" + targetShardRouting.state() + "], expecting to be [initializing]");
         }
 
-        RecoverySourceHandler handler = ongoingRecoveries.addNewRecovery(request, targetShardRouting.allocationId().getId(), shard);
+        // the following IF condition won't go into master, it only goes into 5.4 / 5.x branch to allow for interoperability with < 5.4
+        if (request.targetAllocationId() == null) {
+            // ES versions < 5.4.0 do not send targetAllocationId as part of recovery request, just assume that we have the correct id
+            request = new StartRecoveryRequest(request.shardId(), targetShardRouting.allocationId().getId(), request.sourceNode(),
+                request.targetNode(), request.metadataSnapshot(), request.isPrimaryRelocation(), request.recoveryId(),
+                request.startingSeqNo());
+        }
+
+        if (request.targetAllocationId().equals(targetShardRouting.allocationId().getId()) == false) {
+            logger.debug("delaying recovery of {} due to target allocation id mismatch (expected: [{}], but was: [{}])",
+                request.shardId(), request.targetAllocationId(), targetShardRouting.allocationId().getId());
+            throw new DelayRecoveryException("source node has the state of the target shard to have allocation id [" +
+                targetShardRouting.allocationId().getId() + "], expecting to be [" + request.targetAllocationId() + "]");
+        }
+
+        RecoverySourceHandler handler = ongoingRecoveries.addNewRecovery(request, shard);
         logger.trace("[{}][{}] starting recovery to {}", request.shardId().getIndex().getName(), request.shardId().id(), request.targetNode());
         try {
             return handler.recoverToTarget();
@@ -133,9 +148,9 @@ public class PeerRecoverySourceService extends AbstractComponent implements Inde
     private final class OngoingRecoveries {
         private final Map<IndexShard, ShardRecoveryContext> ongoingRecoveries = new HashMap<>();
 
-        synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, String targetAllocationId, IndexShard shard) {
+        synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
             final ShardRecoveryContext shardContext = ongoingRecoveries.computeIfAbsent(shard, s -> new ShardRecoveryContext());
-            RecoverySourceHandler handler = shardContext.addNewRecovery(request, targetAllocationId, shard);
+            RecoverySourceHandler handler = shardContext.addNewRecovery(request, shard);
             shard.recoveryStats().incCurrentAsSource();
             return handler;
         }
@@ -181,20 +196,19 @@ public class PeerRecoverySourceService extends AbstractComponent implements Inde
              * Adds recovery source handler if recoveries are not delayed from starting (see also {@link #delayNewRecoveries}.
              * Throws {@link DelayRecoveryException} if new recoveries are delayed from starting.
              */
-            synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, String targetAllocationId, IndexShard shard) {
+            synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
                 if (onNewRecoveryException != null) {
                     throw onNewRecoveryException;
                 }
-                RecoverySourceHandler handler = createRecoverySourceHandler(request, targetAllocationId, shard);
+                RecoverySourceHandler handler = createRecoverySourceHandler(request, shard);
                 recoveryHandlers.add(handler);
                 return handler;
             }
 
-            private RecoverySourceHandler createRecoverySourceHandler(StartRecoveryRequest request, String targetAllocationId,
-                                                                      IndexShard shard) {
+            private RecoverySourceHandler createRecoverySourceHandler(StartRecoveryRequest request, IndexShard shard) {
                 RecoverySourceHandler handler;
                 final RemoteRecoveryTargetHandler recoveryTarget =
-                    new RemoteRecoveryTargetHandler(request.recoveryId(), request.shardId(), targetAllocationId, transportService,
+                    new RemoteRecoveryTargetHandler(request.recoveryId(), request.shardId(), transportService,
                         request.targetNode(), recoverySettings, throttleTime -> shard.recoveryStats().addThrottleTime(throttleTime));
                 Supplier<Long> currentClusterStateVersionSupplier = () -> clusterService.state().getVersion();
                 handler = new RecoverySourceHandler(shard, recoveryTarget, request, currentClusterStateVersionSupplier,

--- a/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java
@@ -84,7 +84,7 @@ public class PeerRecoverySourceService extends AbstractComponent implements Inde
         }
     }
 
-    private RecoveryResponse recover(StartRecoveryRequest request) throws IOException {
+    private RecoveryResponse recover(final StartRecoveryRequest request) throws IOException {
         final IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
         final IndexShard shard = indexService.getShard(request.shardId().id());
 
@@ -111,14 +111,6 @@ public class PeerRecoverySourceService extends AbstractComponent implements Inde
             logger.debug("delaying recovery of {} as it is not listed as initializing on the target node {}. known shards state is [{}]",
                 request.shardId(), request.targetNode(), targetShardRouting.state());
             throw new DelayRecoveryException("source node has the state of the target shard to be [" + targetShardRouting.state() + "], expecting to be [initializing]");
-        }
-
-        // the following IF condition won't go into master, it only goes into 5.4 / 5.x branch to allow for interoperability with < 5.4
-        if (request.targetAllocationId() == null) {
-            // ES versions < 5.4.0 do not send targetAllocationId as part of recovery request, just assume that we have the correct id
-            request = new StartRecoveryRequest(request.shardId(), targetShardRouting.allocationId().getId(), request.sourceNode(),
-                request.targetNode(), request.metadataSnapshot(), request.isPrimaryRelocation(), request.recoveryId(),
-                request.startingSeqNo());
         }
 
         if (request.targetAllocationId().equals(targetShardRouting.allocationId().getId()) == false) {

--- a/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java
@@ -329,6 +329,7 @@ public class PeerRecoveryTargetService extends AbstractComponent implements Inde
 
         request = new StartRecoveryRequest(
             recoveryTarget.shardId(),
+            recoveryTarget.indexShard().routingEntry().allocationId().getId(),
             recoveryTarget.sourceNode(),
             clusterService.localNode(),
             metadataSnapshot,

--- a/core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
@@ -78,7 +78,6 @@ public class RecoverySourceHandler {
     protected final Logger logger;
     // Shard that is going to be recovered (the "source")
     private final IndexShard shard;
-    private final String indexName;
     private final int shardId;
     // Request containing source and target node information
     private final StartRecoveryRequest request;
@@ -116,7 +115,6 @@ public class RecoverySourceHandler {
         this.request = request;
         this.currentClusterStateVersionSupplier = currentClusterStateVersionSupplier;
         this.delayNewRecoveries = delayNewRecoveries;
-        this.indexName = this.request.shardId().getIndex().getName();
         this.shardId = this.request.shardId().id();
         this.logger = Loggers.getLogger(getClass(), nodeSettings, request.shardId(), "recover to " + request.targetNode().getName());
         this.chunkSizeInBytes = fileChunkSizeInBytes;
@@ -443,7 +441,7 @@ public class RecoverySourceHandler {
         StopWatch stopWatch = new StopWatch().start();
         logger.trace("finalizing recovery");
         cancellableThreads.execute(() -> {
-            shard.markAllocationIdAsInSync(recoveryTarget.getTargetAllocationId());
+            shard.markAllocationIdAsInSync(request.targetAllocationId());
             recoveryTarget.finalizeRecovery(shard.getGlobalCheckpoint());
         });
 

--- a/core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java
@@ -369,11 +369,6 @@ public class RecoveryTarget extends AbstractRefCounted implements RecoveryTarget
     }
 
     @Override
-    public String getTargetAllocationId() {
-        return indexShard().routingEntry().allocationId().getId();
-    }
-
-    @Override
     public void ensureClusterStateVersion(long clusterStateVersion) {
         ensureClusterStateVersionCallback.handle(clusterStateVersion);
     }

--- a/core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetHandler.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetHandler.java
@@ -77,9 +77,4 @@ public interface RecoveryTargetHandler {
     void writeFileChunk(StoreFileMetaData fileMetaData, long position, BytesReference content,
                         boolean lastChunk, int totalTranslogOps) throws IOException;
 
-    /***
-     * @return the allocation id of the target shard.
-     */
-    String getTargetAllocationId();
-
 }

--- a/core/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java
@@ -50,14 +50,10 @@ public class RemoteRecoveryTargetHandler implements RecoveryTargetHandler {
     private final AtomicLong bytesSinceLastPause = new AtomicLong();
 
     private final Consumer<Long> onSourceThrottle;
-    private String targetAllocationId;
 
-    public RemoteRecoveryTargetHandler(long recoveryId, ShardId shardId, String targetAllocationId, TransportService transportService,
+    public RemoteRecoveryTargetHandler(long recoveryId, ShardId shardId, TransportService transportService,
                                        DiscoveryNode targetNode, RecoverySettings recoverySettings, Consumer<Long> onSourceThrottle) {
-        this.targetAllocationId = targetAllocationId;
         this.transportService = transportService;
-
-
         this.recoveryId = recoveryId;
         this.shardId = shardId;
         this.targetNode = targetNode;
@@ -164,8 +160,4 @@ public class RemoteRecoveryTargetHandler implements RecoveryTargetHandler {
                         throttleTimeInNanos), fileChunkRequestOptions, EmptyTransportResponseHandler.INSTANCE_SAME).txGet();
     }
 
-    @Override
-    public String getTargetAllocationId() {
-        return targetAllocationId;
-    }
 }

--- a/core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java
@@ -39,6 +39,7 @@ public class StartRecoveryRequest extends TransportRequest {
 
     private long recoveryId;
     private ShardId shardId;
+    private String targetAllocationId;
     private DiscoveryNode sourceNode;
     private DiscoveryNode targetNode;
     private Store.MetadataSnapshot metadataSnapshot;
@@ -51,15 +52,17 @@ public class StartRecoveryRequest extends TransportRequest {
     /**
      * Construct a request for starting a peer recovery.
      *
-     * @param shardId           the shard ID to recover
-     * @param sourceNode        the source node to remover from
-     * @param targetNode        the target node to recover to
-     * @param metadataSnapshot  the Lucene metadata
-     * @param primaryRelocation whether or not the recovery is a primary relocation
-     * @param recoveryId        the recovery ID
-     * @param startingSeqNo     the starting sequence number
+     * @param shardId            the shard ID to recover
+     * @param targetAllocationId the allocation id of the target shard
+     * @param sourceNode         the source node to remover from
+     * @param targetNode         the target node to recover to
+     * @param metadataSnapshot   the Lucene metadata
+     * @param primaryRelocation  whether or not the recovery is a primary relocation
+     * @param recoveryId         the recovery ID
+     * @param startingSeqNo      the starting sequence number
      */
     public StartRecoveryRequest(final ShardId shardId,
+                                final String targetAllocationId,
                                 final DiscoveryNode sourceNode,
                                 final DiscoveryNode targetNode,
                                 final Store.MetadataSnapshot metadataSnapshot,
@@ -68,6 +71,7 @@ public class StartRecoveryRequest extends TransportRequest {
                                 final long startingSeqNo) {
         this.recoveryId = recoveryId;
         this.shardId = shardId;
+        this.targetAllocationId = targetAllocationId;
         this.sourceNode = sourceNode;
         this.targetNode = targetNode;
         this.metadataSnapshot = metadataSnapshot;
@@ -81,6 +85,10 @@ public class StartRecoveryRequest extends TransportRequest {
 
     public ShardId shardId() {
         return shardId;
+    }
+
+    public String targetAllocationId() {
+        return targetAllocationId;
     }
 
     public DiscoveryNode sourceNode() {
@@ -108,6 +116,11 @@ public class StartRecoveryRequest extends TransportRequest {
         super.readFrom(in);
         recoveryId = in.readLong();
         shardId = ShardId.readShardId(in);
+        if (in.getVersion().onOrAfter(Version.V_5_4_0_UNRELEASED)) {
+            targetAllocationId = in.readString();
+        } else {
+            targetAllocationId = null;
+        }
         sourceNode = new DiscoveryNode(in);
         targetNode = new DiscoveryNode(in);
         metadataSnapshot = new Store.MetadataSnapshot(in);
@@ -124,6 +137,9 @@ public class StartRecoveryRequest extends TransportRequest {
         super.writeTo(out);
         out.writeLong(recoveryId);
         shardId.writeTo(out);
+        if (out.getVersion().onOrAfter(Version.V_5_4_0_UNRELEASED)) {
+            out.writeString(targetAllocationId);
+        }
         sourceNode.writeTo(out);
         targetNode.writeTo(out);
         metadataSnapshot.writeTo(out);

--- a/core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java
+++ b/core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java
@@ -21,10 +21,8 @@ package org.elasticsearch.indices.recovery;
 
 import org.elasticsearch.Version;
 import org.elasticsearch.cluster.node.DiscoveryNode;
-import org.elasticsearch.cluster.routing.RecoverySource;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
-import org.elasticsearch.index.seqno.SequenceNumbers;
 import org.elasticsearch.index.seqno.SequenceNumbersService;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.index.store.Store;
@@ -116,11 +114,7 @@ public class StartRecoveryRequest extends TransportRequest {
         super.readFrom(in);
         recoveryId = in.readLong();
         shardId = ShardId.readShardId(in);
-        if (in.getVersion().onOrAfter(Version.V_5_4_0_UNRELEASED)) {
-            targetAllocationId = in.readString();
-        } else {
-            targetAllocationId = null;
-        }
+        targetAllocationId = in.readString();
         sourceNode = new DiscoveryNode(in);
         targetNode = new DiscoveryNode(in);
         metadataSnapshot = new Store.MetadataSnapshot(in);
@@ -137,9 +131,7 @@ public class StartRecoveryRequest extends TransportRequest {
         super.writeTo(out);
         out.writeLong(recoveryId);
         shardId.writeTo(out);
-        if (out.getVersion().onOrAfter(Version.V_5_4_0_UNRELEASED)) {
-            out.writeString(targetAllocationId);
-        }
+        out.writeString(targetAllocationId);
         sourceNode.writeTo(out);
         targetNode.writeTo(out);
         metadataSnapshot.writeTo(out);

--- a/core/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java
+++ b/core/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java
@@ -99,6 +99,7 @@ public class RecoverySourceHandlerTests extends ESTestCase {
         final RecoverySettings recoverySettings = new RecoverySettings(settings, service);
         final StartRecoveryRequest request = new StartRecoveryRequest(
             shardId,
+            null,
             new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
             new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
             null,
@@ -155,6 +156,7 @@ public class RecoverySourceHandlerTests extends ESTestCase {
         final long startingSeqNo = randomBoolean() ? SequenceNumbersService.UNASSIGNED_SEQ_NO : randomIntBetween(0, 16);
         final StartRecoveryRequest request = new StartRecoveryRequest(
             shardId,
+            null,
             new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
             new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
             null,
@@ -223,6 +225,7 @@ public class RecoverySourceHandlerTests extends ESTestCase {
         final StartRecoveryRequest request =
             new StartRecoveryRequest(
                 shardId,
+                null,
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 null,
@@ -292,6 +295,7 @@ public class RecoverySourceHandlerTests extends ESTestCase {
         final StartRecoveryRequest request =
             new StartRecoveryRequest(
                 shardId,
+                null,
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 null,
@@ -358,6 +362,7 @@ public class RecoverySourceHandlerTests extends ESTestCase {
         final StartRecoveryRequest request =
             new StartRecoveryRequest(
                 shardId,
+                null,
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 null,
@@ -417,6 +422,7 @@ public class RecoverySourceHandlerTests extends ESTestCase {
         final StartRecoveryRequest request =
             new StartRecoveryRequest(
                 shardId,
+                null,
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT),
                 null,

--- a/core/src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTests.java
+++ b/core/src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTests.java
@@ -65,11 +65,7 @@ public class StartRecoveryRequestTests extends ESTestCase {
         inRequest.readFrom(in);
 
         assertThat(outRequest.shardId(), equalTo(inRequest.shardId()));
-        if (targetNodeVersion.onOrAfter(Version.V_5_4_0_UNRELEASED)) {
-            assertThat(outRequest.targetAllocationId(), equalTo(inRequest.targetAllocationId()));
-        } else {
-            assertNull(inRequest.targetAllocationId());
-        }
+        assertThat(outRequest.targetAllocationId(), equalTo(inRequest.targetAllocationId()));
         assertThat(outRequest.sourceNode(), equalTo(inRequest.sourceNode()));
         assertThat(outRequest.targetNode(), equalTo(inRequest.targetNode()));
         assertThat(outRequest.metadataSnapshot().asMap(), equalTo(inRequest.metadataSnapshot().asMap()));

--- a/core/src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTests.java
+++ b/core/src/test/java/org/elasticsearch/indices/recovery/StartRecoveryRequestTests.java
@@ -21,6 +21,7 @@ package org.elasticsearch.indices.recovery;
 
 import org.elasticsearch.Version;
 import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.UUIDs;
 import org.elasticsearch.common.io.stream.InputStreamStreamInput;
 import org.elasticsearch.common.io.stream.OutputStreamStreamOutput;
 import org.elasticsearch.index.seqno.SequenceNumbers;
@@ -44,6 +45,7 @@ public class StartRecoveryRequestTests extends ESTestCase {
         final Version targetNodeVersion = randomVersion(random());
         final StartRecoveryRequest outRequest = new StartRecoveryRequest(
                 new ShardId("test", "_na_", 0),
+                UUIDs.base64UUID(),
                 new DiscoveryNode("a", buildNewFakeTransportAddress(), emptyMap(), emptySet(), targetNodeVersion),
                 new DiscoveryNode("b", buildNewFakeTransportAddress(), emptyMap(), emptySet(), targetNodeVersion),
                 Store.MetadataSnapshot.EMPTY,
@@ -63,6 +65,11 @@ public class StartRecoveryRequestTests extends ESTestCase {
         inRequest.readFrom(in);
 
         assertThat(outRequest.shardId(), equalTo(inRequest.shardId()));
+        if (targetNodeVersion.onOrAfter(Version.V_5_4_0_UNRELEASED)) {
+            assertThat(outRequest.targetAllocationId(), equalTo(inRequest.targetAllocationId()));
+        } else {
+            assertNull(inRequest.targetAllocationId());
+        }
         assertThat(outRequest.sourceNode(), equalTo(inRequest.sourceNode()));
         assertThat(outRequest.targetNode(), equalTo(inRequest.targetNode()));
         assertThat(outRequest.metadataSnapshot().asMap(), equalTo(inRequest.metadataSnapshot().asMap()));

--- a/test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java
@@ -405,6 +405,7 @@ public abstract class IndexShardTestCase extends ESTestCase {
         }
         replica.prepareForIndexRecovery();
         final RecoveryTarget recoveryTarget = targetSupplier.apply(replica, pNode);
+        final String targetAllocationId = recoveryTarget.indexShard().routingEntry().allocationId().getId();
 
         final Store.MetadataSnapshot snapshot = getMetadataSnapshotOrEmpty(replica);
         final long startingSeqNo;
@@ -414,8 +415,8 @@ public abstract class IndexShardTestCase extends ESTestCase {
             startingSeqNo = SequenceNumbersService.UNASSIGNED_SEQ_NO;
         }
 
-        final StartRecoveryRequest request =
-            new StartRecoveryRequest(replica.shardId(), pNode, rNode, snapshot, false, 0, startingSeqNo);
+        final StartRecoveryRequest request = new StartRecoveryRequest(replica.shardId(), targetAllocationId,
+            pNode, rNode, snapshot, false, 0, startingSeqNo);
         final RecoverySourceHandler recovery = new RecoverySourceHandler(
             primary,
             recoveryTarget,

